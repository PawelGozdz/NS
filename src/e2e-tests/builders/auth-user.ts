import { Kysely, Transaction } from 'kysely';

import { TableNames, TestingE2EFunctions, dialect, kyselyPlugins } from '@app/core';
import { IDatabaseModels } from '@libs/common';
import { AuthenticationServer } from '@libs/testing';

import { UserSeedBuilder } from './user-builder';

const tablesInvolved = [TableNames.USERS, TableNames.AUTH_USERS, TableNames.USER_PROFILES];

const authenticationServer = new AuthenticationServer();

type IDatabaseDaos = IDatabaseModels;

const dbConnection = new Kysely<IDatabaseDaos>({
  dialect,
  plugins: kyselyPlugins,
});
const dbUtils = new TestingE2EFunctions(dbConnection);

export const getAccessToken = () => authenticationServer.generateAccessToken();
export const getRefreshToken = () => authenticationServer.generateRefreshToken();

export const getCookies = () =>
  authenticationServer.getTokensAsCookie({
    accessToken: getAccessToken(),
    refreshToken: getRefreshToken(),
  });

export async function loginUser(transaction?: Transaction<IDatabaseDaos>) {
  if (transaction) {
    return seed(transaction);
  }
  return dbConnection.transaction().execute(async (trx) => seed(trx));
}

async function seed(trx: Transaction<IDatabaseDaos>) {
  await dbUtils.truncateTables(tablesInvolved, trx);

  const seedBuilder = await UserSeedBuilder.create(trx);
  seedBuilder.withUser().withAuthUser().withProfile();
  await seedBuilder.build();
  return seedBuilder;
}
