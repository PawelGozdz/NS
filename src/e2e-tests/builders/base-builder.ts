/* eslint-disable @typescript-eslint/strict-boolean-expressions */
/* eslint-disable @typescript-eslint/prefer-readonly */
/* eslint-disable @typescript-eslint/no-explicit-any */
// src/seed-builders/base-seed-builder.ts
import { Kysely, SelectExpression, sql, Transaction } from 'kysely';
import { InsertExpression } from 'kysely/dist/cjs/parser/insert-values-parser';
import { chunk, sampleSize, take } from 'lodash';

import { IDatabaseModels, TableNames } from '@app/core';

import { GetOptions } from './interfaces';

type GetDaosOptions = {
  limit?: number;
  random?: boolean;
} & ({ first?: true; last?: false } | { first?: false; last?: true } | { first?: false; last?: false });

export abstract class BaseSeedBuilder<T> {
  protected dbConnection: Kysely<IDatabaseModels> | Transaction<IDatabaseModels>;

  protected tableName: TableNames;

  protected daos: T[] = [];

  constructor(dbConnection: Kysely<IDatabaseModels> | Transaction<IDatabaseModels>, tableName: TableNames) {
    this.dbConnection = dbConnection;
    this.tableName = tableName;
  }

  getDaos(options?: GetDaosOptions): T[] {
    let result = this.daos;

    if (options?.random) {
      result = sampleSize(result, options.limit ?? result.length);
    } else if (options?.first) {
      result = take(result, options.limit ?? 1);
    } else if (options?.last) {
      result = take(result.reverse(), options.limit ?? 1);
    } else if (options?.limit) {
      result = take(result, options.limit);
    }

    return result;
  }

  public getTableName() {
    return this.tableName;
  }

  async getRandomIndex(options?: GetOptions): Promise<number | string> {
    const indexes = await this.selectFields({
      ...options,
      limit: 1,
      shuffle: true,
    });
    if (indexes.length === 0) {
      throw new Error(`No records found in table ${this.getTableName()}`);
    }
    return indexes[0].id;
  }

  async getFirst(options?: Pick<GetOptions, 'fields' | 'sort'>): Promise<T> {
    const sort = {
      sortColumn: options?.sort?.sortColumn ?? 'id',
      sortDirection: 'asc' as const,
    };

    const indexes = await this.selectFields({
      limit: 1,
      shuffle: false,
      sort,
    });
    if (indexes.length === 0) {
      throw new Error(`No records found in table ${this.getTableName()}`);
    }
    return indexes[0];
  }

  async getLast(options?: Pick<GetOptions, 'fields' | 'sort'>): Promise<T> {
    const sort = {
      sortColumn: options?.sort?.sortColumn ?? 'id',
      sortDirection: 'desc' as const,
    };

    const indexes = await this.selectFields({
      limit: 1,
      sort,
    });
    if (indexes.length === 0) {
      throw new Error(`No records found in table ${this.getTableName()}`);
    }

    return indexes[0];
  }

  async getMany(qty: number, options?: GetOptions): Promise<T[]> {
    const indexes = await this.selectFields({ ...options, limit: qty });

    if (indexes.length === 0) {
      throw new Error(`No records found in table ${this.getTableName()}`);
    }

    let result = indexes;

    if (options?.stretchRandom) {
      result = Array.from({ length: qty }, () => result[Math.floor(Math.random() * result.length)]);
    } else if (options?.stretch && result.length < qty) {
      result = Array.from({ length: qty }, (_, i) => result[i % result.length]);
    }

    return result;
  }

  async selectFields(options: GetOptions): Promise<any[]> {
    const fields = options?.fields ? options.fields : ['id'];

    let query = this.dbConnection.selectFrom(this.getTableName()).select(fields as unknown as SelectExpression<IDatabaseModels, TableNames>);

    if (options?.shuffle) {
      query = query.orderBy(sql`RANDOM()`);
    }

    if (options?.sort) {
      query = query.orderBy(options.sort.sortColumn, options.sort.sortDirection);
    }

    if (options?.limit) {
      query = query.limit(options.limit);
    }
    return query.execute();
  }

  async insertDaos(values: T[]) {
    const models: T[] = [];

    for await (const chunks of chunk(values, 10000)) {
      models.push(...(await this.insertSmallChunk(chunks as InsertExpression<IDatabaseModels, TableNames>)));
    }

    return models;
  }

  private async insertSmallChunk(values: InsertExpression<IDatabaseModels, TableNames>) {
    return (await this.dbConnection.insertInto(this.getTableName()).values(values).returningAll().execute()) as T[];
  }
}
