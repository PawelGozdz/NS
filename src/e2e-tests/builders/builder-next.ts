import { Database } from '@app/database';
import { IDatabaseModels } from '@libs/common';
import { Transaction, sql } from 'kysely';

interface RelationConfig {
	foreignKey: string;
	primaryKey: string;
	parentEntity: string;
}

interface SeederConfig {
	table: string;
	data: (parent: any, index: number) => Record<string, any>;
	quantity?: number;
	return?: string[] | ((parent: any) => string[]);
	order?: number;
	relations?: { config: SeederConfig; relation: RelationConfig }[];
}

export class DataSeeder {
	constructor(private db: Database) {}

	async seed(configs: SeederConfig[]): Promise<void> {
		configs.sort((a, b) => (a.order || 0) - (b.order || 0));
		for (const config of configs) {
			await this.db.transaction().execute(async (transaction) => {
				await this.seedRecursive(config, transaction);
			});
		}
	}

	private async seedRecursive(config: SeederConfig, transaction: Transaction<IDatabaseModels>, parent?: any, index?: number) {
		console.log('PAREMT', await parent, config);
		const data = Array(config.quantity || 1)
			.fill(0)
			.map((_, i) => (config.data as (parent: any, index: number) => Record<string, any>)(parent, i));

		const returnFields = typeof config.return === 'function' ? config.return(parent) : config.return;

		if (parent) {
			// w childernie data jes toromisem i nie filtruje indeined
			console.log('DATA', data);
		}

		const filteredData = await Promise.all(this.filterUndefinedProperties(data));
		console.log('FILTERED DATA', filteredData);

		const inserted = await transaction
			.insertInto(config.table as keyof IDatabaseModels)
			.values(filteredData as any)
			.returning(returnFields as any)
			.execute();

		console.log('INSERT', inserted);
		console.log('CONFIG RLEATIOSN____________________________', config.relations);
		if (config.relations) {
			for (const { config: relationConfig, relation } of config.relations) {
				if (relation.parentEntity !== config.table) {
					throw new Error(`Invalid relation: ${relation.parentEntity} is not the parent of ${relationConfig.table}`);
				}
				if ((relationConfig.order || 0) <= (config.order || 0)) {
					throw new Error(
						`Invalid order: The order of ${relationConfig.table} cannot be less than or equal to the order of its parent ${config.table}`,
					);
				}
				for (let i = 0; i < inserted.length; i++) {
					// { id: 'a6185a9f-8873-4f1b-b630-3729318bc636', email: 'abc@co' }
					const buildParent = inserted[i];
					// AuthUser dao
					let relationData = relationConfig.data(buildParent, i);
					// console.log('RELATIOAN DATA', relationData);

					relationData[relation.foreignKey] = buildParent[relation.primaryKey];
					// console.log('REL DATA', relationData);
					// relationConfig.data = relationData;
					if (parent) {
						console.log(relationConfig, buildParent, i);
					}
					await this.seedRecursive(relationConfig, transaction, buildParent, i);
				}
			}
		}
		// console.log('INESRE', inserted);

		return inserted;
	}

	async rollback(configs: SeederConfig[]): Promise<void> {
		configs.sort((a, b) => (b.order || 0) - (a.order || 0));
		for (const config of configs) {
			await this.db.transaction().execute(async (transaction) => {
				await this.rollbackRecursive(config, transaction);
			});
		}
	}

	private async rollbackRecursive(config: SeederConfig, transaction: Transaction<IDatabaseModels>): Promise<void> {
		if (config.relations) {
			for (const { config: relationConfig } of config.relations) {
				await this.rollbackRecursive(relationConfig, transaction);
			}
		}

		await sql`DELETE FRIM ${config.table}`.execute(transaction);
	}

	private filterUndefinedProperties(array: Record<string, any>[]) {
		return array.map((obj) => {
			const filteredObj = {};
			Object.entries(obj).forEach(([key, value]) => {
				if (value !== undefined) {
					filteredObj[key] = value;
				}
			});
			return filteredObj;
		});
	}
}
