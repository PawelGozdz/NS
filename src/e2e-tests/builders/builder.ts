import { AuthUserModel, ProfileModel, UserModel } from '@app/contexts/auth';
import { Kysely } from 'kysely';
import { TableNames, dialect, kyselyPlugins } from '../../database';
import { AuthUserFixtureFactory, ProfileFixtureFactory, UserFixtureFactory } from '../fixtures';

type IDatabaseDaos = any;

export class UserSeedBuilder {
	public dbConnection: Kysely<IDatabaseDaos>;
	public userDao: UserModel;
	public authUserDao: AuthUserModel;
	public profileDao: ProfileModel;

	daos: {
		userDaoObj: UserModel | undefined;
		authUserDaoObj: AuthUserModel | undefined;
		profileDaoObj: ProfileModel | undefined;
	} = {
		userDaoObj: undefined,
		authUserDaoObj: undefined,
		profileDaoObj: undefined,
	};

	actions: { method: string }[] = [];
	cookies: string[] = [];

	private constructor(dbConnection: Kysely<IDatabaseDaos>) {
		this.dbConnection = dbConnection;
	}

	async build() {
		for await (const action of this.actions) {
			await this[action.method]();
		}

		return this;
	}

	static async create(db?: Kysely<IDatabaseDaos>): Promise<UserSeedBuilder> {
		const builder = new UserSeedBuilder(
			db ||
				new Kysely<IDatabaseDaos>({
					dialect,
					plugins: kyselyPlugins,
				}),
		);

		return builder;
	}

	private async insertUser() {
		if (!this.daos.userDaoObj) {
			throw new Error('UserDao is not defined');
		}

		await this.dbConnection.insertInto(TableNames.USERS).values(this.daos.userDaoObj).execute();
		this.userDao = (await this.dbConnection
			.selectFrom(TableNames.USERS)
			.selectAll()
			.where('id', '=', this.daos.userDaoObj.id)
			.executeTakeFirst()) as UserModel;
	}

	withUser(user?: Partial<UserModel>): this {
		this.daos.userDaoObj = UserFixtureFactory.create({
			...user,
		});

		this.actions.push({ method: 'insertUser' });

		return this;
	}

	private async insertAuthUser(): Promise<this> {
		if (!this.daos.authUserDaoObj) {
			throw new Error('AuthUserDao is not defined');
		}

		await this.dbConnection.insertInto(TableNames.AUTH_USERS).values(this.daos.authUserDaoObj).execute();
		this.authUserDao = (await this.dbConnection
			.selectFrom(TableNames.AUTH_USERS)
			.selectAll()
			.where('userId', '=', this.daos.authUserDaoObj.userId)
			.executeTakeFirst()) as AuthUserModel;

		return this;
	}

	withAuthUser(authUser?: Partial<AuthUserModel>): this {
		if (!this.daos.userDaoObj) {
			throw new Error('UserDao is not defined');
		}

		this.daos.authUserDaoObj = AuthUserFixtureFactory.create({
			userId: this.daos.userDaoObj.id,
			...authUser,
		});

		this.actions.push({ method: 'insertAuthUser' });

		return this;
	}

	private async insertProfile() {
		if (!this.daos.profileDaoObj || !this.daos.userDaoObj) {
			throw new Error('UserDao is not defined');
		}

		await this.dbConnection.insertInto(TableNames.PROFILES).values(this.daos.profileDaoObj).execute();
		this.profileDao = (await this.dbConnection
			.selectFrom(TableNames.PROFILES)
			.selectAll()
			.where('id', '=', this.daos.profileDaoObj.id)
			.executeTakeFirst()) as ProfileModel;
	}

	withProfile(profile?: Partial<ProfileModel>): this {
		this.daos.profileDaoObj = ProfileFixtureFactory.create({
			userId: this.daos.userDaoObj?.id,
			...profile,
		});

		this.actions.push({ method: 'insertProfile' });

		return this;
	}
}
