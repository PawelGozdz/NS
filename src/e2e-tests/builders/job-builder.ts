/* eslint-disable no-restricted-syntax */
import { Kysely, Transaction } from 'kysely';

import { JobModel, JobPositionModel, JobUserProfileModel, UserProfileModel } from '@app/contexts';
import { IDatabaseModels, TableNames, dialect, kyselyPlugins } from '@app/core';

import { JobFixtureFactory, JobPositionFixtureFactory, JobUserProfileFixtureFactory } from '../fixtures/job.fixture';

type IDatabaseDaos = IDatabaseModels;

export class JobSeedBuilder {
  public dbConnection: Kysely<IDatabaseDaos>;

  public jobDao: JobModel;

  public jobUserProfileDao: JobUserProfileModel;

  public jobPositionDao: JobPositionModel;

  daos: {
    jobDaoObj: JobModel | undefined;
    jobUserProfileDaoObj: JobUserProfileModel | undefined;
    jobPositionDaoObj: JobPositionModel | undefined;
  } = { jobDaoObj: undefined, jobUserProfileDaoObj: undefined, jobPositionDaoObj: undefined };

  actions: { method: string }[] = [];

  cookies: string[] = [];

  private constructor(dbConnection: Kysely<IDatabaseDaos>) {
    this.dbConnection = dbConnection;
  }

  async build() {
    for await (const action of this.actions) {
      await this[action.method]();
    }

    return this;
  }

  static async create(db?: Transaction<IDatabaseDaos>): Promise<JobSeedBuilder> {
    const builder = new JobSeedBuilder(
      db ??
        new Kysely<IDatabaseDaos>({
          dialect,
          plugins: kyselyPlugins,
        }),
    );

    return builder;
  }

  private async insertJob() {
    if (!this.daos.jobDaoObj) {
      throw new Error('JobDao is not defined');
    }

    if (!this.daos.jobPositionDaoObj) {
      throw new Error('JobPositionDao is not defined to insert Job');
    }

    this.jobDao = (await this.dbConnection
      .insertInto(TableNames.JOBS)
      .values({
        ...this.daos.jobDaoObj,
        jobPositionId: this.daos.jobPositionDaoObj.id,
      })
      .returningAll()
      .executeTakeFirst()) as JobModel;
  }

  withJob(dao?: Partial<JobModel>): this {
    this.daos.jobDaoObj = JobFixtureFactory.create({
      ...dao,
    });

    this.actions.push({ method: 'insertJob' });

    return this;
  }

  private async insertJobUserProfile(): Promise<this> {
    if (!this.daos.jobUserProfileDaoObj) {
      throw new Error('JobUserProfile is not defined');
    }

    this.jobUserProfileDao = (await this.dbConnection
      .insertInto(TableNames.JOB_USER_PROFILES)
      .values(this.daos.jobUserProfileDaoObj)
      .returningAll()
      .executeTakeFirst()) as JobUserProfileModel;

    return this;
  }

  withJobUserProfile(entity: Partial<JobUserProfileModel> & { userId: string }): this {
    if (!this.daos.jobUserProfileDaoObj) {
      throw new Error('JobUserProfileDaoObj is not defined');
    }

    this.daos.jobUserProfileDaoObj = JobUserProfileFixtureFactory.create({
      ...entity,
      userId: entity.userId,
    });

    this.actions.push({ method: 'insertJobUserProfile' });

    return this;
  }

  private async insertJobPosition() {
    if (!this.daos.jobPositionDaoObj) {
      throw new Error('JobPositionDao is not defined');
    }

    this.jobPositionDao = (await this.dbConnection
      .insertInto(TableNames.JOB_POSITIONS)
      .values(this.daos.jobPositionDaoObj)
      .returningAll()
      .executeTakeFirst()) as JobPositionModel;
  }

  withJobPosition(profile: Partial<UserProfileModel> & { categoryId: number }): this {
    this.daos.jobPositionDaoObj = JobPositionFixtureFactory.create({
      ...profile,
      categoryId: profile.categoryId,
    });

    this.actions.push({ method: 'insertJobPosition' });

    return this;
  }
}
