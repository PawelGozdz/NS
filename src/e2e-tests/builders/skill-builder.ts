import { Kysely } from 'kysely';

import { CategoryModel, SkillModel } from '@app/contexts';
import { IDatabaseModels, TableNames, dialect, kyselyPlugins } from '@app/core';

import { CategoryFixtureFactory, SkillFixtureFactory } from '../fixtures';

type IDatabaseDaos = IDatabaseModels;

export class SkillSeedBuilder {
  public dbConnection: Kysely<IDatabaseDaos>;

  public skillDao: SkillModel;

  public categoryDao: CategoryModel;

  daos: {
    skillDaoObj: SkillModel | undefined;
    categoryDaoObj: CategoryModel | undefined;
  } = { skillDaoObj: undefined, categoryDaoObj: undefined };

  actions: { method: string }[] = [];

  private constructor(dbConnection: Kysely<IDatabaseDaos>) {
    this.dbConnection = dbConnection;
  }

  async build() {
    // eslint-disable-next-line no-restricted-syntax
    for await (const action of this.actions) {
      await this[action.method]();
    }

    this.actions = [];

    return this;
  }

  static async create(db?: Kysely<IDatabaseDaos>): Promise<SkillSeedBuilder> {
    const builder = new SkillSeedBuilder(
      db ??
        new Kysely<IDatabaseDaos>({
          dialect,
          plugins: kyselyPlugins,
        }),
    );

    return builder;
  }

  private async insertSkill() {
    if (!this.daos.skillDaoObj) {
      throw new Error('skillDaoObj is not defined');
    }

    this.skillDao = (await this.dbConnection
      .insertInto(TableNames.SKILLS)
      .values(this.daos.skillDaoObj)
      .returningAll()
      .executeTakeFirst()) as SkillModel;
  }

  private async insertCategory() {
    if (!this.daos.categoryDaoObj) {
      throw new Error('categoryDaoObj is not defined');
    }

    this.categoryDao = (await this.dbConnection
      .insertInto(TableNames.CATEGORIES)
      .values(this.daos.categoryDaoObj)
      .returningAll()
      .executeTakeFirst()) as CategoryModel;

    if (this.daos.skillDaoObj) {
      this.daos.skillDaoObj.categoryId = this.categoryDao.id;
    }
  }

  withSkill(model?: Partial<SkillModel>): this {
    this.daos.skillDaoObj = SkillFixtureFactory.create({
      ...model,
      categoryId: this.daos.categoryDaoObj?.id,
    });

    this.actions.push({ method: 'insertSkill' });

    return this;
  }

  withCategory(model?: Partial<CategoryModel>): this {
    this.daos.categoryDaoObj = CategoryFixtureFactory.create({
      ...model,
    });

    this.actions.push({ method: 'insertCategory' });

    return this;
  }
}
