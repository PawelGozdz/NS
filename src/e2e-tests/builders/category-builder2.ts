/* eslint-disable no-plusplus */
import { Kysely, Transaction } from 'kysely';

import { CategoryModel } from '@app/contexts';
import { IDatabaseModels, TableNames } from '@app/core';

import { CategoryFixtureFactory } from '../fixtures';
import { BaseSeedBuilder } from './base-builder';
import { InsertOptions } from './interfaces';

export class CategorySeedBuilder2 extends BaseSeedBuilder<CategoryModel> {
  constructor(dbConnection: Kysely<IDatabaseModels> | Transaction<IDatabaseModels>) {
    super(dbConnection, TableNames.CATEGORIES);
  }

  async insert(options: InsertOptions): Promise<void> {
    const quantity = options.quantity ?? 1;

    const values: CategoryModel[] = [];
    for (let i = 0; i < quantity; i++) {
      values.push(CategoryFixtureFactory.createRandom());
    }

    this.daos = await this.insertDaos(values);
  }
}
