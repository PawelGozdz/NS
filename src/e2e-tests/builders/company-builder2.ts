/* eslint-disable no-plusplus */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Kysely, Transaction } from 'kysely';

import { CompanyModel } from '@app/contexts';
import { IDatabaseModels, TableNames } from '@app/core';

import { CompanyFixtureFactory } from '../fixtures';
import { BaseSeedBuilder } from './base-builder';
import { InsertOptions } from './interfaces';

export class CompanySeedBuilder2 extends BaseSeedBuilder<CompanyModel> {
  constructor(dbConnection: Kysely<IDatabaseModels> | Transaction<IDatabaseModels>) {
    super(dbConnection, TableNames.COMPANIES);
  }

  async insert(options: InsertOptions): Promise<void> {
    const quantity = options.quantity ?? 1;

    const values: CompanyModel[] = [];
    for (let i = 0; i < quantity; i++) {
      values.push(CompanyFixtureFactory.createRandom());
    }

    this.daos = await this.insertDaos(values);
  }
}
