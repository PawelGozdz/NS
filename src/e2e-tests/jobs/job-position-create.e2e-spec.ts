import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { Kysely } from 'kysely';
import request from 'supertest';

import { IDatabaseModels, TableNames, TestingE2EFunctions, dialect, kyselyPlugins } from '@app/core';

import { AppModule } from '../../app.module';
import { getCookies, loginUser } from '../builders/auth-user';
import { CategorySeedBuilder } from '../builders/category-builder';
import { JobSeedBuilder } from '../builders/job-builder';
import { SkillSeedBuilder } from '../builders/skill-builder';
import { JobPositionFixtureFactory } from '../fixtures/job.fixture';

type IDdbDaos = IDatabaseModels;

describe('JobsControllerV1 -> create (e2e)', () => {
  const dbConnection = new Kysely<IDdbDaos>({
    dialect,
    plugins: kyselyPlugins,
  });
  const dbUtils = new TestingE2EFunctions(dbConnection);
  let app: INestApplication;

  const tablesInvolved = [TableNames.CATEGORIES, TableNames.SKILLS, TableNames.JOB_POSITIONS, TableNames.JOBS];

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  afterAll(async () => {
    await dbConnection.destroy();
    await app.close();
  });

  let credentials: [string, string];
  let skillBuilder: SkillSeedBuilder;
  let skillBuilder2: SkillSeedBuilder;

  beforeEach(async () => {
    credentials = getCookies();

    await dbConnection.transaction().execute(async (trx) => {
      await dbUtils.truncateTables(tablesInvolved, trx);
      await loginUser(trx);

      skillBuilder = await SkillSeedBuilder.create(trx);
      skillBuilder
        .withCategory({
          name: 'IT',
          description: 'IT category',
        })
        .withSkill({
          name: 'Backend',
          description: 'Backend programming',
        });
      await skillBuilder.build();

      skillBuilder2 = await SkillSeedBuilder.create(trx);
      skillBuilder2
        .withCategory({
          name: 'IT',
          description: 'IT category',
        })
        .withSkill({
          name: 'HR',
          description: 'Researcher',
        });
      await skillBuilder2.build();
    });
  });

  describe('/jos-positions (POST) V1', () => {
    describe('SUCCESS', () => {
      it('should create job and return id', async () => {
        // Arrange
        const entity = JobPositionFixtureFactory.create();

        // Act
        const response = await request(app.getHttpServer())
          .post('/job-positions')
          .set(...credentials)
          .set('Content-Type', 'application/json')
          .send({
            ...entity,
            skillIds: [skillBuilder.skillDao.id, skillBuilder2.skillDao.id],
            categoryId: skillBuilder.categoryDao.id,
          });

        const inserted = await dbConnection
          .selectFrom(TableNames.JOB_POSITIONS)
          .selectAll()
          .where('id', '=', response.body.data.id)
          .executeTakeFirst();

        // Assert
        expect(response.statusCode).toBe(201);
        expect(response.body.data).toEqual({
          id: inserted?.id,
        });
        expect(inserted).toMatchSnapshot({
          title: entity.title,
          slug: entity.slug,
          categoryId: expect.any(Number),
          skillIds: expect.any(Array),
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
          id: expect.any(String),
        });
      });

      it('should create job with existing title but different categoryId', async () => {
        // Arrange
        const entity = JobPositionFixtureFactory.create({
          title: 'Backend Developer',
        });

        const categoryBuilder = await CategorySeedBuilder.create(dbConnection);
        categoryBuilder.withCategory({
          name: 'IT',
          description: 'IT category',
        });
        await categoryBuilder.build();

        const jobPositionBuilder = await JobSeedBuilder.create(dbConnection);
        jobPositionBuilder.withJobPosition({
          ...entity,
          categoryId: categoryBuilder.categoryDao.id,
        });
        await jobPositionBuilder.build();

        // Act
        const response = await request(app.getHttpServer())
          .post('/job-positions')
          .set(...credentials)
          .set('Content-Type', 'application/json')
          .send({
            ...entity,
            skillIds: [skillBuilder.skillDao.id],
            categoryId: skillBuilder.categoryDao.id,
          });

        const inserted = await dbConnection
          .selectFrom(TableNames.JOB_POSITIONS)
          .selectAll()
          .where('id', '=', response.body.data.id)
          .executeTakeFirst();

        // Assert
        expect(response.statusCode).toBe(201);
        expect(response.body.data).toEqual({
          id: inserted?.id,
        });
        expect(inserted).toMatchSnapshot({
          title: entity.title,
          slug: entity.slug,
          categoryId: expect.any(Number),
          skillIds: expect.any(Array),
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
          id: expect.any(String),
        });
      });
    });

    describe('FAILURE', () => {
      it('should throw 409 for combination title + categoryId', async () => {
        // Arrange
        const entity = JobPositionFixtureFactory.create({
          title: 'Backend Developer',
        });

        const categoryBuilder = await CategorySeedBuilder.create(dbConnection);
        categoryBuilder.withCategory({
          name: 'IT',
          description: 'IT category',
        });
        await categoryBuilder.build();

        const jobPositionBuilder = await JobSeedBuilder.create(dbConnection);
        jobPositionBuilder.withJobPosition({
          ...entity,
          categoryId: categoryBuilder.categoryDao.id,
        });
        await jobPositionBuilder.build();

        // Act
        const response = await request(app.getHttpServer())
          .post('/job-positions')
          .set(...credentials)
          .set('Content-Type', 'application/json')
          .send({
            ...entity,
            skillIds: [skillBuilder.skillDao.id],
            categoryId: categoryBuilder.categoryDao.id,
          });

        // Assert
        expect(response.statusCode).toBe(409);
        expect(response.body.data).toMatchSnapshot({
          error: expect.stringContaining('Entity with title'),
        });
      });

      it('should return error if given skill id no exists', async () => {
        // Arrange
        const entity = JobPositionFixtureFactory.create();

        // Act
        const response = await request(app.getHttpServer())
          .post('/job-positions')
          .set(...credentials)
          .set('Content-Type', 'application/json')
          .send({
            ...entity,
            skillIds: [skillBuilder.skillDao.id, 1, 2, 3],
            categoryId: skillBuilder.categoryDao.id,
          });

        // Assert
        // expect(response.statusCode).toBe(400);
        expect(response.body.data).toMatchSnapshot();
      });
    });
  });
});
