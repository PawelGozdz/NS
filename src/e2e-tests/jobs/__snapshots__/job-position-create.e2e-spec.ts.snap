// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobsControllerV1 -> create (e2e) /jos-positions (POST) V1 FAILURE should return error if given skill id no exists 1`] = `
{
  "error": "Requested skill ids [1,2,3] don't exists",
}
`;

exports[`JobsControllerV1 -> create (e2e) /jos-positions (POST) V1 FAILURE should throw 409 for combination title + categoryId 1`] = `
{
  "error": StringContaining "Entity with title",
}
`;

exports[`JobsControllerV1 -> create (e2e) /jos-positions (POST) V1 SUCCESS should create job and return id 1`] = `
{
  "categoryId": Any<Number>,
  "createdAt": Any<Date>,
  "id": Any<String>,
  "skillIds": Any<Array>,
  "slug": "test-job-position",
  "title": "Test job position",
  "updatedAt": Any<Date>,
}
`;

exports[`JobsControllerV1 -> create (e2e) /jos-positions (POST) V1 SUCCESS should create job with existing title but different categoryId 1`] = `
{
  "categoryId": Any<Number>,
  "createdAt": Any<Date>,
  "id": Any<String>,
  "skillIds": Any<Array>,
  "slug": "backend-developer",
  "title": "Backend Developer",
  "updatedAt": Any<Date>,
}
`;
